### ISSUE TITLE
[Bug]: colorbar.remove() does not always work

### ISSUE THREAD

**Issue by sylvain-bougnoux** [matplotlib/matplotlib#22257]
### Bug summary

I am trying to use a filled colorbar ([as issued in ](https://stackoverflow.com/questions/44498631/continuous-colorbar-with-contour-levels/44499443#44499443)). But then `cb.remove()` does not work as it used to do, ie the space used by the colorbar is not recovered.

Using `cb.remove()` does work as expected when the colobar is constructed with the output of `plt.contour` as in the initial demo.

### Code for reproduction

```python
import numpy as np
import matplotlib.colors
import matplotlib.pyplot as plt

delta = 0.025
x = np.arange(-3.0, 3.0, delta)
y = np.arange(-2.0, 2.0, delta)
X, Y = np.meshgrid(x, x)
Z1 = np.exp(-X**2 - Y**2)
Z2 = np.exp(-(X - 1)**2 - (Y - 1)**2)
Z = (Z1 - Z2) * 2

fig, ax = plt.subplots()

cs = plt.contour(X, Y, Z, cmap="viridis")

norm= matplotlib.colors.Normalize(vmin=cs.cvalues.min(), vmax=cs.cvalues.max())
sm = plt.cm.ScalarMappable(norm=norm, cmap = cs.cmap)

cb=fig.colorbar(sm, ticks=cs.levels) # bugged, cb.remove() incomplete
#cb=fig.colorbar(cs, ticks=cs.levels) # ok, cb.remove() ok

plt.show()

####
# USE cb.remove() after in an interactive environment
####
```

### Actual outcome

Initial

![python_2022-01-18_11-57-01](https://user-images.githubusercontent.com/6706662/149924455-dd2a33b1-1a06-4d53-8b85-044d8f818eed.png)

After calling `cb.remove()` the space used by the colorbar is not recovered

![python_2022-01-18_11-57-29](https://user-images.githubusercontent.com/6706662/149924533-66c0ae30-660c-476f-a41c-3801432d70b1.png)

### Expected outcome

Expected output (obtained using `cs` instead of `sm`):

![python_2022-01-18_11-58-56](https://user-images.githubusercontent.com/6706662/149924730-5084d7cf-a922-4a0a-9bf2-f6c5570c7322.png)

### Operating system

Windows

### Matplotlib Version

3.5.1

### Matplotlib Backend

Qt5Agg

### Python version

3.8.9

### Installation

pip

**Comment by sylvain-bougnoux** [comment]
Thanks for your prompt answer. Actually I don't know.
If you have a fast workaround, for some reasons resetting the axes position hardly work either.

**Comment by jklymak** [comment]
I don't think this ever worked.   We get the info about the axes to resize from the mappable.  I suppose its possible we could get that info another way, but thats the way it works now, so adding this would be a feature request.

**Comment by sylvain-bougnoux** [comment]
In case of inconsistent API, the frontier between missing feature & bug is vague :-)

FYI the fast workaround is
```
cb.remove()
ax.set_position([...])  # left,bottom,width,height
ax.set_subplotspec(GridSpec(1,1)[0:1,0:1])
```

**Comment by anntzer** [comment]
From a quick check this seems reasonably easy to fix: currently Colorbar.remove tries to unsteal space from the mappable axes
```
        try:
            ax = self.mappable.axes
        except AttributeError:
            return
        try:
            gs = ax.get_subplotspec().get_gridspec()
            subplotspec = gs.get_topmost_subplotspec()
        except AttributeError:
            # use_gridspec was False
            pos = ax.get_position(original=True)
            ax._set_position(pos)
        else:
            # use_gridspec was True
            ax.set_subplotspec(subplotspec)
```
but actually we know the actual axes from which space was stolen: it's the axes in `cb.ax._colorbar_info["parents"]`.  I *suspect* the fix is to simply unsteal from each of these axes instead.

**Comment by timhoffm** [comment]
Required action: https://github.com/matplotlib/matplotlib/issues/22257#issuecomment-1332798619 should be tried
